### PR TITLE
[BD-18] Switch codejail from nose to pytest

### PR DESCRIPTION
**Pull request by tehreem-sadat**
**Description**:

- Replace nose testing framework with pytest.

**Packages removed**
nose

**New packages**
pytest

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @tehreem-sadat! I've created [BLENDED-396](https://openedx.atlassian.net/browse/BLENDED-396) to keep track of it in Jira.

When this pull request is ready, tag your edX technical lead.

**Comment by morenol** [comment]
Hi @tehreem-sadat could you run `make upgrade` on top of this changes? It seems that this was run with `pip-tools==4.0.0`

**Comment by tehreem-sadat** [comment]
@morenol I have updated changes.

**Comment by feanil** [comment]
jenkins run python 3.8

**Comment by feanil** [comment]
jenkins test python 3.5

**Comment by feanil** [comment]
jenkins test python 3.5

**Comment by awais786** [comment]
jenkins test python 3.5

**Comment by awais786** [comment]
jenkins ok to test

**Comment by edx-webhook** [comment]
Hello @tehreem-sadat: We are unable to continue with review of your submission at this time. Please see the associated JIRA ticket for more explanation.

**Comment by morenol** [comment]
Hi @tehreem-sadat, I think that it is failing due to virtualenv, could you add this to constraints.txt and then run make upgrade again?

virtualenv<20.0.0

**Comment by morenol** [comment]
@tehreem-sadat you can ignore the python3.8 tests, I am working on that in the another PR. They never have passed

**Comment by tehreem-sadat** [comment]
@morenol I have updated codejail/jail_code.py, with this latest change we don't need virtualenv>=20.0.0 check. Please have a look at it.

**Comment by morenol** [comment]
Cool, it looks good to me. My only concern is that according to [this](https://stackoverflow.com/a/58026969).

> The problem is, this only works when the environment is activated by the activate shell script.
> 
> You can start the environment's scripts without activating the environment, so if that is a concern, you have to use a different method.
> 

I am not completely aware of all the use cases of codejail, is it possible that we run something from that environment without setting that virtualenv?

Maybe we can try, the other option suggested in that post: 
```python
import sys

real_prefix = getattr(sys, "real_prefix", None)
base_prefix = getattr(sys, "base_prefix", sys.prefix)

running_in_virtualenv = (base_prefix or real_prefix) != sys.prefix
```

**Comment by feanil** [comment]
@tehreem-sadat I think @morenol makes a good point.  There are ocassions where we do just call the relevant python directly so I think we should account for that if it's not too difficult.  Can you see if his suggestion here will work, and if not, we can land as is with some comments to help people debug in the future.

**Comment by tehreem-sadat** [comment]
@morenol, @feanil it worked, thanks for your suggestions. I have updated the changes.

File: codejail/jail_code.py [hunk 1]
```diff
@@ -62,7 +62,10 @@ def is_configured(command):
 
 # By default, look where our current Python is, and maybe there's a
 # python-sandbox alongside.  Only do this if running in a virtualenv.
-if hasattr(sys, 'real_prefix'):
+real_prefix = getattr(sys, "real_prefix", None)
```
Comment (feanil) [inline codejail/jail_code.py:65]: @tehreem-sadat I think this change has caused some dowstream issues in edx-platform and had to be reverted when we got there. @morenol is looking to see if the issue persists on older versions of the virtualenv package but we may need to change or revert some part of this.

Comment (feanil) [reply codejail/jail_code.py:65]: No action needed at this time, but just informing you.  If we need to change something I'll let you know.

**Pull request merged**